src/derivative_free.jl

```diff
@@ -119,471 +32,534 @@ end
 
-function steffensen_itr{T<:AbstractFloat}(f, x0::T;
-                   xtol=4*eps(), xtolrel=4*eps(), ftol=4*eps(),
-                   maxsteps=100, maxfnevals=100)
-    update = (o) -> begin
-        xn = o.xn[end]
-        fxn = o.fxn[end]
-        
-        wn = xn + steff_step(xn, fxn)
-        fwn = f(wn)
-        incfn(o)
-        
-        fp, issue = _fbracket(xn, wn, fxn, fwn)
-        
-        if issue
-            o.state = :converged
-            return
-        end
-        
-        xn1 = xn - fxn / fp
-        
-        
-        fxn1 = o.f(xn1)
-        incfn(o)
-        
-        push!(o.xn, xn1)
-        push!(o.fxn, fxn1)
-
+## order 0
+# goal: more robust to initial guess than higher order methods
+# follows roughly algorithm described http://www.hpl.hp.com/hpjournal/pdfs/IssuePDFs/1979-12.pdf, the SOLVE button from the HP-34C
+# though some modifications were made. (Much faster as older `SOLVE`, but not quite as robust.)
+# * use secant step
+# * if along the way a bracket is found, switch to bisection. (We use float64 bisection not a42 if available)
+# * if secant step fails to decrease, we use quadratic step up to 3 times
+# 
+# Goal is to return a value `x` with either:
+# * `f(x) == 0.0` or 
+# * `f(prevfloat(x)) * f(nextfloat(x)) < 0`.
+# if a bracket is found that can be done, otherwise secant step is used
+function update_state{T}(method::Order0, fs, o::UnivariateZeroState{T}, options::UnivariateZeroOptions)
+
+    f = fs.f
+    alpha, beta = o.xn0, o.xn1
+    falpha, fbeta = o.fxn0, o.fxn1
+    S = eltype(falpha)
+    
+    incsteps(o)
+
+    if sign(falpha) * sign(fbeta) < 0.0
+        # use bisection
+        opts = deepcopy(options); opts.verbose=false
+        find_zero(Bisection(), fs, o, opts)
+        o.message = "Used bisection for last step: [a,b] = [$alpha, $beta]"
+        return nothing
     end
 
-    x = float(x0)
-    x, fx = promote(x, f(x))
-    out = ZeroType(f, nothing, nothing, update, [x], [fx],
-                   xtol, xtolrel, ftol, :not_converged, 
-                   0, maxsteps, 1, maxfnevals)
-
-    out
-end
-
-
-## http://www.naturalspublishing.com/files/published/ahb21733nf19a5.pdf
-## A New Fifth Order Derivative Free Newton-Type Method for Solving Nonlinear Equations
-## Manoj Kumar, Akhilesh Kumar Singh, and Akanksha Srivastava
-## Appl. Math. Inf. Sci. 9, No. 3, 1507-1513 (2015)
-function kss5_itr(f, x0::Real;
-             xtol=4*eps(), xtolrel=4*eps(), ftol=4*eps(),
-             maxsteps=100, maxfnevals=100)
+    gamma, issue = guarded_secant_step(alpha, beta, falpha, fbeta)
+    if issue
+        o.message = "error with guarded secant step"
+        o.stopped = true
+        return nothing
+    end
 
-    update = o -> begin
-        xn = o.xn[end]
-        fxn = o.fxn[end]
+    fgamma = f(gamma); incfn(o)
+    if sign(fgamma)*sign(fbeta) < 0.0
+        o.xn0, o.xn1 = gamma, beta
+        o.fxn0, o.fxn1 = fgamma, fbeta
+        opts = deepcopy(options); opts.verbose=false
```

Comment:
`deepcopy` is notoriously slow. Perhaps better to define a custom `copy` method for `UnivariateZeroOptions`.